### ISSUE TITLE
Lightweight Isolates & Faster isolate communication

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#36097]
This issue is tracking our ongoing work on making isolates more lightweight and improving communication between them. This work will mostly focus on isolates which were spawned from the same source.

As part of this work we will investigate a number of approaches, including:

  - [x] Faster sending of mutable messages across isolates
  - [x] Sharing of program structures (e.g. libraries, classes, functions, fields), JITed code, ...
  - [x] Allowing multiple mutator threads to run in parallel the same code (yet using separate global state)

We would like to implement this in several stages, here's our current progress:

  - [x] [done] Change embedder API to allow creation and initialization of isolate groups
  - [x] [done] Add support for standalone/flutter embedders to initialize isolate-group spawnees
  - [x] [done] Enable multiple isolates to share one heap.
  - [x] [done] Split up existing field objects into mutable and immutable part.
  - [x] [done] Share program structure in AOT mode.
  - [x] [done] Implement `SendPort.sendAndExit` to allow close to 0-cost sending of messages
  - [x] [done] Migrate embedders to take advantage of isolate groups.
  - [x] Sharing JITed code: Will require making major parts of VM re-entrant from multiple mutators

**Subtasks for AOT enable by-default**

  - [x] Ensure all existing embedders that use `Isolate.spawn()` API have been augmented to supply `Dart_InitializeParams.initialize_isolate` callback with correct implementation.

**Subtasks for JIT**

  - [x] Reduce `RunWithMutatorsStopped` in SwitchableCall miss handler to only around patching the tuple in the pool (and remove hack in symbol table)
  - [x] Enable class finalization on BG compiler by adding appropriate locking: Allows us to test this in normal JIT mode and is pre-requisite for many mutator threads
  - [x] Make shared BG compiler queue that multiple mutators can use (i.e. move BG compiler functionality from isolate to the group)

Future work tracked at [Issue 46752](https://github.com/dart-lang/sdk/issues/46752)

**NOTICE**

This work will require substantial restructuring of the Dart VM internals and is therefore relatively risky and giving a concrete timeline is not possible - that being said we expect to make good progress on this in the coming months.

The (old) isolates will continue to be available as they were before and we will try to ensure to not regress any existing code.

What the actual improvements in terms of isolate spawning time, parallelization as well as communications will be can only really be seen once the implementation is done. That being said, we can get numbers for some features in an intermediate stage.

**Comment by mkustermann** [comment]
Thread local allocation buffers (TLABs) have been restored in df34f65d90, 5737b452b and the background compiler has been changed to allow allocating into new space in fb88c983e7.

**Comment by mkustermann** [comment]
A prototype for faster sending of mutable messages across isolates has been made in [cl/92280](https://dart-review.googlesource.com/c/sdk/+/92280), which might allow 2-3x faster serialization and almost constant deserialization time.

**Comment by mkustermann** [comment]
A prototype for transferrable byte arrays is available in [cl/86640](https://dart-review.googlesource.com/c/sdk/+/86640)

**Comment by mkustermann** [comment]
Before we can start working on heap unification we have to change the isolate spawning functionality which currently just calls out to the embedder and assumes the embedder happens to load the exact same code into the isolate. See https://github.com/dart-lang/sdk/issues/36648 (which is blocking)

**Comment by aam** [comment]
transferable typeddata landed in https://dart-review.googlesource.com/c/sdk/+/99623

**Comment by xster** [comment]
Eagerly looking for the outcome of this feature request :D 

Considering you guys seem to have made a lot of progress on this already, do you have a design doc outlining the general intended outcome or usage of the new paradigm?

**Comment by shafqatevo** [comment]
It is understood that lot more work is still to be done in this direction. Is it still possible to share how lightweight isolates may become? In terms of what parameters the lightweightness is aimed?

There are some references in the web that Dart isolates have a memory overhead of 2MB, which is more in the range of OS thread overhead. Is the goal of this work to make isolates lightweight like green threads? There are also references in the web that the original design goal of Dart isolates was to follow actor model - is that still the direction?

For green threads or lightweight process concurrency, comparisons would be memory overheads of Kotlin’s coroutines (“a few dozen bytes”), Akka actors (~300 bytes), Pony actors (256 bytes), Erlang processes (309 words = 309x8 bytes in 64-bit architectures).

**Comment by mkustermann** [comment]
> It is understood that lot more work is still to be done in this direction. Is it still possible to share how lightweight isolates may 
> become? In terms of what parameters the lightweightness is aimed?
> There are some references in the web that Dart isolates have a memory overhead of 2MB, which is more in the range of OS 
> thread overhead. Is the goal of this work to make isolates lightweight like green threads?

There are several goals we have:

* Faster `Isolate.spawn()` latency: Spawning isolates right now takes time to spawn that is linear in the application size. We would like to bring it closer to be a constant time overhead.
* Less memory per additional isolate: Spawning an isolate consumes right now non-user memory that is linear in application size. We would like to bring it closer to use constant non-user memory overhead per additional isolate.
 
Furthermore we would like to improve communication speed between isolates. Making it faster, smaller and possibly removing some restrictions we currently impose for inter-isolate messages.

> There are also references in the web  that the original design goal of Dart isolates was to follow actor model - is that still the direction?

Isolates are independent from each other, have their own event loop and communicate by sending each other messages - that's how it is today and it will continue to be like that. Though we might investigate possibilities of e.g. sharing immutable memory instead of copying messages when sending.


> Isolates are not for lightweight concurrency as such. But for green threads or lightweight process concurrency, comparisons 
> would be memory overheads of Kotlin’s coroutines (“a few dozen bytes”), Akka actors (~300 bytes), Pony actors (256 bytes), 
> Erlang processes (309 words = 309x8 bytes in 64-bit architectures).

What is "light" depends on the point of reference, which in the context of this issue is the current implementation of isolates in the Dart VM.

We are working hard to reduce the per-isolate memory overhead, though we'll be unlikely to bring down the memory overhead to low single digit KBs.

The programming models are different in various ways: In Dart an isolate has its own event loop, its own global fields, can run in parallel with other isolates and have their own live-cycle. That has benefits and drawbacks to other models which choose to make different trade-offs.

**Comment by maxim-saplin** [comment]
> A prototype for faster sending of mutable messages across isolates has been made in [cl/92280](https://dart-review.googlesource.com/c/sdk/+/92280), which might allow 2-3x faster serialization and close to O(1) deserialization time.

Is there a chance to see that improvement in Dev/Master channel any time soon? While extensively using issolates to off-load heavy lifting I often see severe freezes of the main thread when receving 30-40 MBs back from an isolate (Dart object with several large List<String> collections)

**Comment by mkustermann** [comment]
@maxim-saplin We're prioritising finishing the JIT support atm and will look at communication related tasks afterwards.

**Comment by maks** [comment]
@maxim-saplin I haven't tried it yet myself, but maybe this using [TransferableTypedData](https://api.dart.dev/stable/2.10.5/dart-isolate/TransferableTypedData-class.html) could be of help for transfering large amounts of data between your Isolates?

**Comment by maxim-saplin** [comment]
@maks thanks for the suggestion, but in my case I deal with large List<String>. There's no way to effeciently get bytes of a string as Uint8List and vice versa - I found none. Using Utf8 encoder/decoder is slower than dumb byte copying as there's data tranaformation (one byte string being converted to aither one byte ASCII chars or 2 byte UTF16 chars). Also you can get string's codeUnits as List<int>, and it is possible to create a string from code units, but under the hood there's much processing happening, not a simple byte copy as well.

**Comment by tatumizer** [comment]
@maxim-saplin: FWIW, I have a program (written 8 years ago) that serializes data into Uint8List. Everything is pretty straightforward, except Strings. For Strings, after quite a bit of experimentation, I found out that this variant of the code is the fastest.
```dart
 writeString(value) {
    // optimistically, 1 byte per char
    int firstPos=bufPos;
    _ensureSpace(8+value.length*2);
    writeTypeAndLength(_STRING1, value.length);
    bool isOneByteString=true;
    for (int i=0; i<value.length; i++) {
      var codeUnit=value.codeUnitAt(i);
      buf[bufPos++]= codeUnit;
      if ((codeUnit&0xFF00)!=0) {
        isOneByteString=false;
        break;
      }
    }
    if (isOneByteString)
      return;
    bufPos=firstPos;
    writeTypeAndLength(_STRING2, value.length);
    for (int i=0; i<value.length; i++) {
      var codeUnit=value.codeUnitAt(i);
      buf[bufPos]= codeUnit;
      buf[bufPos+1]= codeUnit>>8;
      bufPos+=2;
    }  
  } 
```

**Comment by dnfield** [comment]
Will lightweight isolates spawn new threads? If so, will it be possible to control the priority of those threads?

**Comment by mkustermann** [comment]
> Will lightweight isolates spawn new threads? If so, will it be possible to control the priority of those threads?

Lightweight isolates spawned by the embedder can be scheduled by the embedder - e.g. flutter does so already (afaik) if it makes multiple lightweight UI isolates.

The VM runs them on a thread pool (if they have no work to do they don't occupy a thread) with a maximum size. The VM doesn't modify thread priorities.

**Comment by dnfield** [comment]
So how would this interact with https://github.com/dart-lang/sdk/issues/44209?

**Comment by xster** [comment]
> Lightweight isolates spawned by the embedder can be scheduled by the embedder - e.g. flutter does so already (afaik) if it makes multiple lightweight UI isolates.

The multi-engine multi-isolate usage currently has all the engines and isolates share the same thread. So its priority is set once when the thread is created (on Android, and never actually set on iOS https://github.com/flutter/flutter/issues/65752)

**Comment by mkustermann** [comment]
> So how would this interact with #44209?

Any isolate spawned via `Isolate.spawn()` is managed by the VM and is run on VM's thread pool. The priorities are not specified for those atm (they get inherited from whatever thread causes the thread pool to create a new thread). If `--worker-thread-priority` is specified that priority is used for any thread spawned by the VM.

Isolates created by the embedder are generally run by the embedder and the embedder therefore determines which threads it uses for it and what priorities those threads have. The only exception is if the embedder passes responsibility of running such an isolate to the VM via `Dart_RunLoopAsync`.

**Comment by mkustermann** [comment]
As part of making it safe to share JITed code we run into some unexpected existing issues with generated code where a background compiler (or other mutator) could cause deoptimization at places where it wasn't safe to do.

Those issues have been fixed as part of https://github.com/dart-lang/sdk/issues/45213#issuecomment-869603807 (see summary [here](https://github.com/dart-lang/sdk/issues/45213#issuecomment-869603807)).

That included these changes:
* [dart-lang/sdk@960cc715ee2](https://github.com/dart-lang/sdk/commit/960cc715ee2)
* [dart-lang/sdk@1bdd9da521d](https://github.com/dart-lang/sdk/commit/1bdd9da521d)
* [dart-lang/sdk@a44d1eefe67](https://github.com/dart-lang/sdk/commit/a44d1eefe67)
* [dart-lang/sdk@e674aa456f1](https://github.com/dart-lang/sdk/commit/e674aa456f1)
* [dart-lang/sdk@535b53659b4](https://github.com/dart-lang/sdk/commit/535b53659b4)
* [dart-lang/sdk@9dee9c77d73](https://github.com/dart-lang/sdk/commit/9dee9c77d73)
* [dart-lang/sdk@a791ecc30e5](https://github.com/dart-lang/sdk/commit/a791ecc30e5)
* [dart-lang/sdk@dce0263228c](https://github.com/dart-lang/sdk/commit/dce0263228c)
* [dart-lang/sdk@67ffc84c05c](https://github.com/dart-lang/sdk/commit/67ffc84c05c)
* [dart-lang/sdk@8a4cba83a27](https://github.com/dart-lang/sdk/commit/8a4cba83a27)
* [dart-lang/sdk@3c81d992ef5](https://github.com/dart-lang/sdk/commit/3c81d992ef5)
* [dart-lang/sdk@de767e3f2d2](https://github.com/dart-lang/sdk/commit/de767e3f2d2)
* [dart-lang/sdk@6127386cfae](https://github.com/dart-lang/sdk/commit/6127386cfae)
* [dart-lang/sdk@6a585a20899](https://github.com/dart-lang/sdk/commit/6a585a20899)
* [dart-lang/sdk@bdb77e4f3d6](https://github.com/dart-lang/sdk/commit/bdb77e4f3d6)
* [dart-lang/sdk@deb946745ca](https://github.com/dart-lang/sdk/commit/deb946745ca)
* [dart-lang/sdk@dec7b4f5b67](https://github.com/dart-lang/sdk/commit/dec7b4f5b67)
* [dart-lang/sdk@caa31f55be5](https://github.com/dart-lang/sdk/commit/caa31f55be5)
* [dart-lang/sdk@dbd2c76e22d](https://github.com/dart-lang/sdk/commit/dbd2c76e22d)
* [dart-lang/sdk@4c0edc67207](https://github.com/dart-lang/sdk/commit/4c0edc67207)
* [dart-lang/sdk@df439941e4c](https://github.com/dart-lang/sdk/commit/df439941e4c)
* [dart-lang/sdk@26f0eba03b7](https://github.com/dart-lang/sdk/commit/26f0eba03b7)
* [dart-lang/sdk@0834f9b7ac8](https://github.com/dart-lang/sdk/commit/0834f9b7ac8)
* [dart-lang/sdk@1055db533c6](https://github.com/dart-lang/sdk/commit/1055db533c6)

**Comment by gmpassos** [comment]
Note that only the ability to have frozen objects (immutable after some copy, or for the entire life of the object), with an optimized way to share them between isolates (without copy/serialize them), is already a very important feature, that can allow a significant improvement in isolate performance. Maybe this feature should be released before.

**Comment by mkustermann** [comment]
@gmpassos What we're about to finish is the foundation on which any object sharing is based on. Without this foundation (sharing program structure, JITed code, ...) the sharing is not possible. Once we're done, transitive object copies will be much faster than they currently are. We'll also share constants and other immutable objects such as strings.

Freezing of object graphs in particular is a little controversial among the team - so it's unclear where this will go. It's speculative at this point.

@gmpassos What would help us a lot is to know more about customer use cases - that makes decision making for us easier. Could you share some more details about how you would benefit from lightweight isolates and object sharing (e.g. what kinds of object you'd like to share, how big those object graphs are, how often messages are exchanged between isolates - i.e. frequency and bandwidth of inter-isolate communication)?

**Comment by gmpassos** [comment]
Well, the easiest way to share something between multiple threads is to have an immutable object graph: no copy, no lock and no bottlenecks.

Since the parallelism paradigm of Dart is "Isolate", the tradeoff is data sharing (coping/serializing/immutables) between threads and the use of parallelism heavily dependent on the Actor model (message-based).

Do not forget that a Turing machine, in his essence, is only transforming data, and the CPU cost and time is directly proportional to the amount of data to process.

In a Actor model parallelism, if the message is the part to be processed, delegated to another Thread/Isolate, there's a tendency that the messaging part (serialize/copy and send the message) to be almost the time to process the message in a single-thread paradigm.

The Actor mode parallelism is only interesting if the message is an instruction, and not the data to be processed, like an URL, or an ID of the part of the data to be processed (like the element index in a List).

The most common data to be processed is a List, that usually is splitted and than multiple threads will process this List, each one processing a sub-set of elements of the List. The biggest issue is how to send this List, or a sub-set, to the other Isole! With immutable objects graphs this problem is reduced to something very simple, a pointer to an immutable object.

I'm the author of the packages:
- https://pub.dev/packages/async_task
- https://pub.dev/packages/async_extension

They help to write portable asynchronous code and to have a good performance.

**Comment by gmpassos** [comment]
Another use case is for ANN.

I'm the author of:
https://pub.dev/packages/eneural_net

This packages focuses to compute ANN using SIMD.

There's the potential to compute the gradient of the ANN (for Backpropagation and similar training algorithms) using multiple threads. I have a current working branch that does that, but the bottleneck is to send the ANN to the Isolate. Actually is better to compute everything in a single-thread, since the time that it takes to send the ANN to another Isolate is just terrible.

**Comment by gmpassos** [comment]
I think that the use of `AsyncTaskChannel ` with `AsyncTask` is what you want.
https://pub.dev/packages/async_task#asynctaskchannel

**Comment by mkustermann** [comment]
We intend to enable this support by-default after [sdk/issues/46754](https://github.com/dart-lang/sdk/issues/46754) was approved.

**Comment by mkustermann** [comment]
This has now landed: Isolate groups are on by-default. Closing this issue.

**Comment by kevmoo** [comment]
Please add an entry to the changelog!

**Comment by dnfield** [comment]
Where is `SendPort.sendAndExit`? I still don't see it available as public API.

**Comment by aam** [comment]
@dnfield there is https://github.com/dart-lang/sdk/issues/47164 that tracks making `sendAndExit` available.
